carma-messenger-core/cpp_message/src/MobilityRequest_Message.cpp

```diff
@@ -0,0 +1,538 @@
+        trajectory_location->timestamp.size=trajectory_timestamp_string_size;
+        message->value.choice.TestMessage00.body.trajectoryStart=trajectory_location;
+
+            //trajectory offsets
+        size_t offset_count=plainMessage.trajectory.offsets.size();
+        
+        if(offset_count>MAX_POINTS_IN_MESSAGE){
+            ROS_WARN_STREAM("offset count greater than 60.");
+            return boost::optional<std::vector<uint8_t>>{};
+        }
+
+        MobilityLocationOffsets* offsets_list;
+        offsets_list=(MobilityLocationOffsets*)calloc(1,sizeof(MobilityLocationOffsets));
+        if(!offsets_list)
+        {
+            ROS_WARN_STREAM("Cannot allocate mem for offsets list encoding");
+            return boost::optional<std::vector<uint8_t>>{};
+        }
+
+        MobilityECEFOffset* Offsets;    
+        for(size_t i=0;i<offset_count;i++){
+            Offsets=(MobilityECEFOffset*)calloc(1,sizeof(MobilityECEFOffset));
+            if(!Offsets)
+            {
+                ROS_WARN_STREAM("Cannot allocate mem for Offsets encoding");
+                return boost::optional<std::vector<uint8_t>>{};
+            }
+            Offsets->offsetX=plainMessage.trajectory.offsets[i].offset_x;
+            Offsets->offsetY=plainMessage.trajectory.offsets[i].offset_y;
+            Offsets->offsetZ=plainMessage.trajectory.offsets[i].offset_z;
+            asn_sequence_add(&offsets_list->list,Offsets);
+        }
+        
+        message->value.choice.TestMessage00.body.trajectory=offsets_list;
+
+        // //expiration
+        uint64_t expiration_message=plainMessage.expiration;
+        std::string expiration_string=std::to_string(expiration_message);
+        size_t expiration_string_size=expiration_string.size();
+        uint8_t expiration_array[expiration_string_size];  
+        for(size_t i=0;i<expiration_string_size ;i++)
+        {
+            expiration_array[i]=expiration_string[i];
+        }
+
+        MobilityTimestamp_t* expiration_time;
+        expiration_time=(MobilityTimestamp_t*)calloc(1,sizeof(MobilityTimestamp_t));
+        if(!expiration_time)
+        {
+            ROS_WARN_STREAM("Cannot allocate mem for expiration message encoding");
+            return boost::optional<std::vector<uint8_t>>{};
+        }
+        expiration_time->size=expiration_string_size;
+        expiration_time->buf=expiration_array;
+        message->value.choice.TestMessage00.body.expiration=expiration_time;
+
+
+        ec=uper_encode_to_buffer(&asn_DEF_MessageFrame,0, message, buffer, buffer_size);
+        if(ec.encoded==-1){;
+            return boost::optional<std::vector<uint8_t>>{}; 
+        }     
+        //copy to byte array msg
+        size_t array_length=ec.encoded/8;
+        std::vector<uint8_t> b_array(array_length);
+        for(size_t i = 0; i < array_length; i++) 
+        {
+            b_array[i] = buffer[i];
+        }
+        return boost::optional<std::vector<uint8_t>>(b_array);
```

Comment:
Big issue here is that I don't see any frees/deletes to free up the allocated memory from the callocs in this function. We're going to be leaking memory every time this method is called. Another complication to add to that is that this method has multiple exit points (whenever we fail to encode, or a value fails validation and you return an empty optional) and we need to make sure that all memory cleanup happens before each of those exit points.

Since this is a complex method than that, the most reliable form of memory management here is likely to be C++ STD smart pointers. I'd recommend taking a look at std::shared_ptr in particular. These smart pointers will handle freeing their memory resources automatically whenever they go out of scope, so you don't need to worry about manually `delete`ing them.